### PR TITLE
Closes #73 improve derive_interp_records

### LINKED ISSUE #73: General Issue: Improve `derive_interp_records()`
### Background Information

- The final dataset shouldn't be assigned to a variable. Then the returned dataset is printed. At the moment we get:
   ```
    > derive_interp_records(my_cdc, by_vars = exprs(SEX), parameter = "HEIGHT")
    > 
    ```
    Then also the example in the documentation could be simplified to
    ```
    library(dplyr, warn.conflicts = FALSE)
    library(rlang)

    admiralpeds::cdc_htage %>%
      mutate(
        SEX = case_when(
          SEX == 1 ~ "M",
          SEX == 2 ~ "F",
          TRUE ~ NA_character_
        ),
        # Ensure first that Age unit is "DAYS"
        AGE = round(AGE * 30.4375)
      ) %>%
      # Interpolate the AGE by SEX
      derive_interp_records(
        by_vars = exprs(SEX),
        parameter = "HEIGHT"
      )
    ```
- The function drops variables which are not expected. `derive_*()` functions shouldn't drop any variables or records.
- Why is the uniqueness of records only checked if `by_vars` is `NULL`? I wonder if the check should be replaced by calling `signal_duplicate_records(dataset, by_vars = exprs(!!!by_vars, AGE))` regardless of the value of `by_vars`.
- In the documentation of the `dataset` argument `AGEU` is listed as required variable. However, this isn't checked nor is the variable used in the code.
- The "Value" section of the documentation should provide more details. I had to read the code to understand what the function is doing. For example it should be stated that _linear_ interpolation is done and that the output dataset contains a record for each day from the minimum age to the maximum age.

### Definition of Done

_No response_

### PR DESCRIPTION
**Pull request by Fanny-Gautier**
Thank you for your Pull Request! We have developed this task checklist from the [Development Process Guide](https://pharmaverse.github.io/admiral/CONTRIBUTING.html#detailed-development-process) to help with the final steps of the process. Completing the below tasks helps to ensure our reviewers can maximize their time on your code as well as making sure the admiral codebase remains robust and consistent.   

Please check off each taskbox as an acknowledgment that you completed the task or check off that it is not relevant to your Pull Request. This checklist is part of the Github Action workflows and the Pull Request will not be merged into the `main` branch until you have checked off each task.

- [x] Place Closes #<insert_issue_number> into the beginning of your Pull Request Title (Use Edit button in top-right if you need to update)
- [x] Code is formatted according to the [tidyverse style guide](https://style.tidyverse.org/). Run `styler::style_file()` to style R and Rmd files
- [x] Updated relevant unit tests or have written new unit tests, which should consider realistic data scenarios and edge cases, e.g. empty datasets, errors, boundary cases etc. - See [Unit Test Guide](https://pharmaverse.github.io/admiraldev/articles/unit_test_guidance.html#tests-should-be-robust-to-cover-realistic-data-scenarios)
- [x] If you removed/replaced any function and/or function parameters, did you fully follow the [deprecation guidance](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#deprecation)?
- [x] Update to all relevant roxygen headers and examples, including keywords and families. Refer to the [categorization of functions](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#categorization-of-functions) to tag appropriate keyword/family.
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the `NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md` under the header `# admiralpeds (development version)` if the changes pertain to a user-facing function (i.e. it has an `@export` tag) or documentation aimed at users (rather than developers)
- [x] Build admiralpeds site `pkgdown::build_site()` and check that all affected examples are displayed correctly and that all new functions occur on the Reference page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue in the Development Section on the right hand side.
- [x] Address all merge conflicts and resolve appropriately
- [x] Pat yourself on the back for a job well done! Much love to your accomplishment!

### CODE REVIEW

**Comment by Fanny-Gautier** [comment]
@bundfussr ,
Could you please suggest how to get ride of the following error ?
image

When I run lintr in my session I don't get any error.
image

It's well added in  `admiralpeds-package.R`, and `NAMESPACE`
image
image
 and Imported in `DESCRIPTION`

image

**Comment by rossfarrugia** [comment]
@Fanny-Gautier  as per discussion on the other PR please also make any vignette updates as a result of these template changes so both remain consistent in the same PR for ease of review, and you'll need to merge latest main branch (resolving any conflicts)

**Comment by Fanny-Gautier** [comment]
> @bundfussr , Could you please suggest how to get ride of the following error ? Thanks. image
> 
> When I run lintr in my session I don't get any error. image
> 
> It's well added in `admiralpeds-package.R`, and `NAMESPACE` image image and Imported in `DESCRIPTION`
> 
> image

Hi @rossfarrugia , could you please suggest how to fixe this error that I can finish this PR and assign the reviewers? thanks

**Comment by rossfarrugia** [comment]
@bundfuss i need your advice for the above please. i tried a few things as you can see in the commits above and don't get why lintr fails here. R CMD check is fine so what Fanny has added seems fine to me.

@Fanny-Gautier in the meantime feel free to add reviewers and mark as ready as it looks to me like an error on the CI side not with anything you've done. We use the same function in admiralroche in same way as how you've added and our gitlab CI doesn't ever have a problem.

**Comment by bundfussr** [comment]
> @bundfuss i need your advice for the above please. i tried a few things as you can see in the commits above and don't get why lintr fails here. R CMD check is fine so what Fanny has added seems fine to me.

@rossfarrugia , yes, that's very strange. Could we clear the cache and rerun the workflow? Unfortunately, I don't have the rights to clear the cache.

**Comment by Fanny-Gautier** [comment]
🎉🎉 Thanks @bundfussr for the solution and update !! Good to know that the dependencies must be defined !

File: R/derive_interp_records.R [hunk 1]
```diff
@@ -95,22 +105,32 @@ derive_interp_records <- function(dataset,
...
         age <- AGE
-        x <- lapply(select(., all_of(metadata_vars)), fapp, age = age)
+        x <- lapply(select(., all_of(interp_vars)), fapp, age = age)
         as.data.frame(do.call(bind_cols, x))
       }) %>%
       filter(!is.na(AGE))
   } else {
-    dataset <- dataset %>%
+    interp_dataset <- dataset %>%
       group_by(across(!!!syms(map(replace_values_by_names(by_vars), as_label)))) %>%
```
Comment (bundfussr) [inline R/derive_interp_records.R:117]:
```suggestion
      group_by(!!!by_vars, AGEU) %>%
```
Then `AGEU` is populated for all records and the `if` branch above can be removed.

File: R/derive_interp_records.R [hunk 2]
```diff
@@ -44,48 +45,57 @@
...
-      ))
-    }
+  # Define the metadata variables to be interpolated
+  interp_vars <- c("AGE", "L", "M", "S")
+  if (parameter == "BMI") {
+    interp_vars <- append(interp_vars, c("P95", "Sigma"))
+  }
+
+  # Define the non-interpolated variables and keep the corresponding unique records
+  non_interp_vars <- setdiff(names(dataset), c(interp_vars, by_vars))
```
Comment (bundfussr) [inline R/derive_interp_records.R:94]:
```suggestion
  non_interp_vars <- setdiff(names(dataset), c(interp_vars, by_vars, expr(AGEU))
```
as `AGEU` should be added to `group_by()`.

**Review by bundfussr — CHANGES_REQUESTED** [review]

File: R/derive_interp_records.R [hunk 3]
```diff
@@ -95,22 +105,32 @@ derive_interp_records <- function(dataset,
...
       filter(!is.na(AGE))
   }
+
+  # Merge non-interpolated variables (if any) back into the interpolated dataset
+  if (length(non_interp_vars) > 0) {
+    final_dataset <- interp_dataset %>%
+      left_join(non_interp_dataset, by = map_chr(replace_values_by_names(by_vars), as_label))
+  } else {
+    final_dataset <- interp_dataset
+  }
```
[the remark this answers is not in the document]
Comment (Fanny-Gautier) [reply R/derive_interp_records.R:133]: The above code is also not working as `AGE` is not in the `non_interp_dataset` dataset.
The problem comes from `signal_duplicate_records(dataset, by_vars = exprs(!!!by_vars, AGE))` which does not check for the uniqueness of the records like in above example. In the above example, `by_vars` should be updated to `by_vars = exprs(SEX,A)`.
The previous code applied to all cases (not only when `by_vars = NULL`) would throw that error and fixe the merging problem in the end:
image

**Comment by Fanny-Gautier** [comment]
Due to system issue (login fails due to worldwide Microsoft issue), I haven't been able to progress much on the new comments. Hoping it's working again on Monday.

File: R/derive_interp_records.R [hunk 4]
```diff
@@ -44,48 +45,57 @@
...
+    filter(toupper(AGEU) != "DAYS") %>%
+    group_by(AGEU) %>%
+    slice(1) %>%
+    ungroup()
 
-  # Ensure to have unique combination when by_vars is not defined
-  metadata_vars <- c("AGE", "L", "M", "S")
-  if (parameter == "BMI") {
-    metadata_vars <- append(metadata_vars, c("P95", "Sigma"))
+  if (nrow(ageu) > 0) {
```
[the remark this answers is not in the document]
Comment (Fanny-Gautier) [reply R/derive_interp_records.R]: implemented with toupper()

File: R/derive_interp_records.R (hunk 3 above)
[the remark this answers is not in the document]
Comment (Fanny-Gautier) [reply R/derive_interp_records.R]: Merged by AGE, and LOCF applied

**Comment by Fanny-Gautier** [comment]
Having system issue again, I raised a Ticket to my IT support... @bundfussr could you please review what I have already committed today and let me know whether this is good enough ? Thanks

File: R/derive_interp_records.R (hunk 1 above)
Comment (Fanny-Gautier) [reply R/derive_interp_records.R]: `!!!by_vars` is not sufficient. If `by_vars` contains several variables it resolves to an error like "Object not found" : at least `map_chr(replace_values_by_names(by_vars), as_label)` is needed here.

File: R/derive_interp_records.R [hunk 5]
```diff
@@ -44,48 +45,57 @@
...
+  if (parameter == "BMI") {
+    interp_vars <- append(interp_vars, c("P95", "Sigma"))
+  }
+
+  # Define the non-interpolated variables and keep the corresponding unique records
+  non_interp_vars <- setdiff(names(dataset), c(interp_vars, by_vars))
+  if (length(non_interp_vars) > 0) {
+    non_interp_dataset <- dataset %>%
+      select(map_chr(replace_values_by_names(by_vars), as_label), all_of(non_interp_vars)) %>%
+      unique()
```
[the remark this answers is not in the document]
Comment (Fanny-Gautier) [reply R/derive_interp_records.R:94]: `non_interp_vars` is not updated as `AGE` is not part of the variables to which LOCF is applied. `unique()` is here still needed.

File: R/derive_interp_records.R [hunk 6]
```diff
@@ -44,48 +45,57 @@
...
+  interp_vars <- c("AGE", "L", "M", "S")
+  if (parameter == "BMI") {
+    interp_vars <- append(interp_vars, c("P95", "Sigma"))
+  }
+
+  # Define the non-interpolated variables and keep the corresponding unique records
+  non_interp_vars <- setdiff(names(dataset), c(interp_vars, by_vars))
+  if (length(non_interp_vars) > 0) {
+    non_interp_dataset <- dataset %>%
+      select(map_chr(replace_values_by_names(by_vars), as_label), all_of(non_interp_vars)) %>%
```
[the remark this answers is not in the document]
Comment (Fanny-Gautier) [reply R/derive_interp_records.R]: `!!!by_vars` is not sufficient. If `by_vars` contains several variables it resolves to an error like "Object not found" : at least `map_chr(replace_values_by_names(by_vars), as_label)` is needed here.
`AGE` has been added to the selection.

File: R/derive_interp_records.R (hunk 2 above)
Comment (Fanny-Gautier) [reply R/derive_interp_records.R:87]: `AGEU` is the result of the difference between variables names of the `dataset` and ( `interp_vars` + `by_vars`). No need to specify it here.

File: R/derive_interp_records.R (hunk 6 above)
[the remark this answers is not in the document]
Comment (bundfussr) [reply R/derive_interp_records.R]: That's strange. We use `select(!!!by_vars)` in many places in the admiral code. I tried the suggested change. All unit tests passed and also with more than one by variable I got no error.

Could you provide an example where it fails?

File: R/derive_interp_records.R (hunk 5 above)
[the remark this answers is not in the document]
Comment (bundfussr) [reply R/derive_interp_records.R:94]: `by_vars` and `AGE` are a unique key of `dataset` and included in `non_interp_dataset`. Thus I would expect that `unique()` has no effect. Or did I miss something?

File: R/derive_interp_records.R [hunk 7]
```diff
@@ -24,7 +24,8 @@
 #'        `parameter = "HEIGHT"`,
 #'   or   `parameter = "BMI"`.
 #'
-#' @return The input dataset plus additional interpolated records.
+#' @return The input dataset plus additional interpolated records: a record
+#'         for each day from the minimum age to the maximum age.
```
Comment (bundfussr) [inline R/derive_interp_records.R:28]: It should be mentioned how the non-expected variables are populated in the output dataset.

File: R/derive_interp_records.R [hunk 8]
```diff
@@ -44,73 +45,102 @@
...
-
-    if (nb_occ > 1) {
-      cli_abort(paste0(
-        "The combination of ", paste(other_data_vars, collapse = ", "),
-        " must be unique. Please define `by_vars` otherwise."
-      ))
-    }
+  # Define the non-interpolated variables and keep the corresponding unique records
+  non_interp_vars <- setdiff(names(dataset), c(interp_vars, by_vars))
+  if (length(non_interp_vars) > 0) {
```
Comment (bundfussr) [inline R/derive_interp_records.R:88]: This condition should be remove because `non_interp_vars` contains at least `AGEU`.

File: R/derive_interp_records.R [hunk 9]
```diff
@@ -44,73 +45,102 @@
...
-        x <- lapply(across(all_of(metadata_vars)), fapp, age = age)
+        x <- lapply(across(all_of(interp_vars)), fapp, age = age)
         as.data.frame(do.call(bind_cols, x))
       }) %>%
       ungroup() %>%
       filter(!is.na(AGE))
   }
+
+  # Merge non-interpolated variables (if any) back into the interpolated dataset
+  if (length(non_interp_vars) > 0) {
```
Comment (bundfussr) [inline R/derive_interp_records.R:129]: This condition and the `else` branch should be remove because `non_interp_vars` contains at least `AGEU`.

**Review by bundfussr — CHANGES_REQUESTED** [review]

**Review by bundfussr — APPROVED** [review]

**Review by rossfarrugia — APPROVED** [review]
